saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java

```diff
@@ -0,0 +1,583 @@
+package edu.illinois.cs.cogcomp.saul.learn;
+
+import edu.illinois.cs.cogcomp.core.datastructures.vectors.ExceptionlessInputStream;
+import edu.illinois.cs.cogcomp.core.datastructures.vectors.ExceptionlessOutputStream;
+import edu.illinois.cs.cogcomp.lbjava.classify.*;
+import edu.illinois.cs.cogcomp.lbjava.learn.Learner;
+
+import java.io.PrintStream;
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Use the <code>MultilayerPerceptron</code> to train and subsequently test a neural network for regression or classification.
+ *
+ * <p>One can specify the number of hidden layers as well as the size of each layer and the activation function to use for each layer.
+ * For classification tasks, it is strongly recommended to use a sigmoid activation at the final layer. This implementation works
+ * for both sparse and dense feature vectors. Use the appropriate functions based on the representation of your feature vector.
+ * </p>
+ *
+ * @author      Abiyaz Chowdhury <chowdh2@cooper.edu>
+ */
+
+final public class MultilayerPerceptron extends Learner{
+    /**
+     * Activation function codes. When calling the {@link MultilayerPerceptron} constructor to specify activation functions, use these values to specify
+     * the activation functions.
+     */
+    private static final int SIGMOID = 1;
+    private static final int TANH = 2;
+    private static final int RELU = 3;
+    private static final int LEAKY_RELU = 4;
+    private static final int SOFTPLUS = 5;
+
+    /**
+     * Default neural network learning parameters.
+     */
+    private static double default_learning_rate = 0.1d;
+    private static double default_momentum = 0.1d;
+    private static double default_lambda = 0.00001d;
+
+    /**
+     * Default parameters for network initialization. For classification tasks, each node of the output layer corresponds
+     * to a class. The default setting below is for binary classification.
+     */
+    private static int default_num_inputs = 128;
+    private static int[] default_hidden_layers =  new int[]{};
+    private static int[] default_activation_function = {SIGMOID};
+    private static int default_num_outputs = 2;
+
+    private static int iterations = 0;
+    private static int iterations2 = 0;
+
+    public Layer[] layers;
+
+    /** Each layer contains a set of weights, gradients, and activation values. The aggregate of all layers comprise
+     * the entire neural network.
+     */
+    public class Layer {
+        public double[][] weights;
+        private double[][] gradients;
+        private double[][] prev_gradients;
+        private double[] pre_activation;
+        private double[] post_activation;
+        private double[] errors;
+        private int activation_function;
+
+        // Initializes a layer of the neural network. Weights are all initialized to 0.
+        private Layer(int num_inputs, int num_outputs, int activation_function){
+            weights = new double[num_inputs+1][num_outputs];
```

Comment:
Minor: I believe primitives in Java are initialized to 0 by default. You might not need to fill the arrays with 0.